src/main/groovy/life/qbic/portal/infrastructure/services/workflowtracking/WorkflowTrackingServiceConnector.groovy

```diff
@@ -1,8 +1,72 @@
 package life.qbic.portal.infrastructure.services.workflowtracking
 
+import groovy.json.JsonSlurper
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.client.RxHttpClient
+import life.qbic.datamodel.workflows.RunInfo
+import life.qbic.datamodel.workflows.Trace
 import life.qbic.portal.domain.listworkflowruns.WorkflowTrackingDataSource
+import life.qbic.portal.utils.ConfigurationManagerFactory
+import life.qbic.services.ConsulServiceFactory
+import life.qbic.services.Service
+import life.qbic.services.ServiceConnector
+import life.qbic.services.ServiceType
+import life.qbic.services.connectors.ConsulConnector
 
-class WorkflowTrackingServiceConnector implements WorkflowTrackingDataSource{
+class WorkflowTrackingServiceConnector implements WorkflowTrackingDataSource {
 
+    URL serviceRegistryURL
+    List<WorkflowTrackingService> workflowTrackingServiceList
 
+    WorkflowTrackingServiceConnector() {
+        this.serviceRegistryURL = loadServiceRegistryURLFromConfig()
```

Comment:
A lot of stuff can go wrong here, and there is no exception handling.

Have a look how I solved it: https://github.com/qbicsoftware/sample-status-updater-cli/blob/master/sample-status-updater-application/src/main/java/life/qbic/samplestatusupdater/app/SampleUpdater.java

The Connector should be instantiated with a Service. It creates to much responsibility for the Connector class to also search for the correct service and handle exceptions. You will only create side effects that are difficult to test. 

You application code should search for a service and create the connector, otherwise it shall fail and notify the user.

Reply:
I did move the search for the service out of the application code by intent to keep the use case separated from the data source which can be basically anything as long as it implements the interface. I will however move the service search out of the connector class and handle it at initialisation in the `DependencyManager`. Would that suffice?